pkg/cmd/init.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2022 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/MakeNowJust/heredoc/v2"
+	"github.com/spf13/cobra"
+	yamlv2 "gopkg.in/yaml.v2"
+
+	"k8c.io/kubeone/examples"
+	kubeonev1beta2 "k8c.io/kubeone/pkg/apis/kubeone/v1beta2"
+	"k8c.io/kubeone/pkg/fail"
+	"k8c.io/kubeone/pkg/yamled"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/yaml"
+)
+
+type initProvider struct {
+	terraformPath string
+	inTree        bool
+	cloudConfig   string
+	csiConfig     string
+}
+
+var (
+	validProviders = map[string]initProvider{
+		"aws": {
+			terraformPath: "terraform/aws",
+		},
+		"azure": {
+			terraformPath: "terraform/azure",
+			cloudConfig: heredoc.Doc(`
+				{
+				    "tenantId": "{{ .Credentials.AZURE_TENANT_ID }}",
+				    "subscriptionId": "{{ .Credentials.AZURE_SUBSCRIPTION_ID }}",
+				    "aadClientId": "{{ .Credentials.AZURE_CLIENT_ID }}",
+				    "aadClientSecret": "{{ .Credentials.AZURE_CLIENT_SECRET }}",
+				    "resourceGroup": "<RESOURCE-GROUP>",
+				    "location": "<LOCATION>",
+				    "subnetName": "<SUBNET>",
+				    "routeTableName": "",
+				    "securityGroupName": "<SECURITY-GROUP>",
+				    "vnetName": "<VPC-NAME>",
+				    "primaryAvailabilitySetName": "<AVAILABILITY-SET-NAME>",
```

Comment:
No, because terraform configurationdoesn't exists yet, nether terraform state.

Reply:
This is true, however, if you enable Terraform generation, we can predict values. For example, you should be able to predict the subnet name, and security group name, because all of those values are generated from the cluster name. You can't predict values such as location, but I think we can try to do as best as we can.

Reply:
Similarly, for vSphere and OpenStack, you can get credentials and other values from the environment.